extensions/notebook/src/dialog/managePackages/managePackagesDialogModel.ts

```diff
@@ -0,0 +1,204 @@
+		return this._packageManageProviders;
+	}
+
+	/**
+	 * Returns the current provider
+	 */
+	public get currentPackageManageProvider(): IPackageManageProvider {
+		if (this.currentProvider) {
+			let provider = this._packageManageProviders.get(this.currentProvider);
+			return provider;
+		}
+		return undefined;
+	}
+
+	/**
+	 * Returns true if multi locations mode is enabled
+	 */
+	public get multiLocationMode(): boolean {
+		return this._options.multiLocations;
+	}
+
+	/**
+	 * returns the array of target locations
+	 */
+	public targetLocations(): string[] {
+		return Array.from(this._locations.keys());
+	}
+
+	/**
+	 * Returns the default location
+	 */
+	public get defaultLocation(): string {
+		return this._options.defaultLocation || this.targetLocations()[0];
+	}
+
+	/**
+	 * Loads the provider cache
+	 */
+	private async loadCaches(): Promise<void> {
+		if (this.packageManageProviders) {
+			let keyArray = Array.from(this.packageManageProviders.keys());
+			for (let index = 0; index < keyArray.length; index++) {
+				const element = this.packageManageProviders.get(keyArray[index]);
+				if (await element.canUseProvider()) {
+					if (!this._locations.has(element.packageTarget.location)) {
+						this._locations.add(element.packageTarget.location);
+					}
+					if (!this._packageTypes.has(element.packageTarget.location)) {
+						this._packageTypes.set(element.packageTarget.location, []);
+					}
+					this._packageTypes.get(element.packageTarget.location).push(element);
+				}
+			}
+		}
+	}
+
+	/**
+	 * Returns the package types for given location
+	 */
+	public getPackageTypes(targetLocation?: string): [string, string][] {
+		targetLocation = targetLocation || this.defaultLocation;
+		let providers = this._packageTypes.get(targetLocation);
+		return providers.map(x => [x.providerId, x.packageTarget.packageType]);
+	}
+
+	/**
+	 * returns the list of packages for current provider
+	 */
+	public async listPackage(): Promise<IPackageDetails[]> {
+		let provider = this.currentPackageManageProvider;
+		if (provider) {
+			return await provider.listPackages();
+		} else {
+			return Promise.reject('Current Provider is not set');
+		}
+	}
+
+	/**
+	 * Changes the current provider
+	 */
+	public changeProvider(providerId: string): void {
+		if (this._packageManageProviders.has(providerId)) {
+			this.currentProvider = providerId;
+		} else {
+			throw Error(`Invalid package type ${providerId}`);
+		}
+	}
+
+	/**
+	 * Installs give packages using current provider
+	 * @param packages Packages to install
+	 */
+	public async installPackages(packages: IPackageDetails[]): Promise<void> {
+		let provider = this.currentPackageManageProvider;
+		if (provider) {
+			await provider.installPackage(packages, false);
+		} else {
+			return Promise.reject('Current Provider is not set');
+		}
+	}
+
+	/**
+	 * Returns the location title for current provider
+	 */
+	public getLocationTitle(): string {
```

Comment:
string | undefined

Reply:
string can be undefined too no? 

---
In reply to: [350545572](https://github.com/microsoft/azuredatastudio/pull/8473#discussion_r350545572) [](ancestors = 350545572)